fauna/client/client.py

```diff
@@ -579,6 +609,50 @@ def _set_endpoint(self, endpoint):
     self._endpoint = endpoint
 
 
+class StreamIterator:
+  """A class that mix a ContextManager and an Iterator so we can detected retryable errors."""
+
+  def __init__(self, client, fql):
+    self.client = client
+    self.fql = fql
+    self.ctx = self.client._stream(self.fql, None)
+    self.stream = None
+    self.last_ts = 0
+    self.error = False
+
+  def __enter__(self):
+    self.stream = self.ctx.__enter__()
+    return self
+
+  def __exit__(self, exc_type, exc_value, exc_traceback):
+    self.ctx.__exit__(exc_type, exc_value, exc_traceback)
+    return True
+
+  def __iter__(self):
+    return self
+
+  def __next__(self):
+    try:
+      if not self.error and self.stream is not None:
+        event: Any = FaunaDecoder.decode(next(self.stream))
+        self.last_ts = event["ts"]
+        self.error = event["type"] == "error"
+        return event
+
+      raise StopIteration
+    except NetworkError as e:
+      return self._retry()
```

Comment:
I'd suggest modifying the Retryable class in this lib and using it here. It accepts a callable, which should allow you to encapsulate all the logic you need here.

You'll have to make the exceptions it retries on are configurable, and update RetryableResponse so it can be more than just QuerySuccess and you should be good.

Then you get to take advantage of two things:
(1) you can configure backoff, which is probably desirable so users don't end up in a tight retry loop during network interruptions
(2) you can configure a max number of retry attempts before we throw, which is also desirable so users can handle longer outages appropriately

Have a look at it. If it doesn't work for this use case for some reason, you'll prob need to implement the above two behaviors.

Reply:
Not sure I will be able to do this because this class is a ContextManager and is meant to be used with `with` clause, but I will try.